experiments/process_sandbox/include/process_sandbox/netpolicy.h

```diff
@@ -0,0 +1,245 @@
+ * use the network must be proxied.
+ */
+
+#pragma once
+#include <netdb.h>
+#include <optional>
+#include <string>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <tuple>
+#include <type_traits>
+#include <unordered_map>
+#include <variant>
+
+namespace sandbox
+{
+  /**
+   * Class encapsulating a network access policy for a sandbox.  This allows
+   * specific access to some network resources to be granted to a sandbox.
+   */
+  class NetworkPolicy
+  {
+    /**
+     * C++17 made noexcept part of the type system, so `R(Args...) noexcept` is
+     * not the same as `R(Args...)`.  Unfortunately, `std::function` remains
+     * specialised only for `R(Args...)` and so you cannot create a
+     * `std::function<R(Args...) noexcept>`.  Fortunately, the `noexcept`
+     * version is a subtype of the potentially throwing version (throwing 0
+     * exceptions is a stronger constraint than throwing 0 or more exceptions)
+     * and so we can just strip `noexcept` from the types.
+     *
+     * This helper provides a mechanism for stripping the `noexcept` qualifier
+     * from a function type.
+     */
+    template<typename Res, typename... ArgTypes>
+    struct remove_noexcept
+    {};
+
+    /**
+     * Specialisation for types that are already not-`noexcept`, just forwards
+     * the type.
+     */
+    template<typename Res, typename... ArgTypes>
+    struct remove_noexcept<Res(ArgTypes...)>
+    {
+      /**
+       * The type (already) without `noexcept`.
+       */
+      using type = Res(ArgTypes...);
+    };
+
+    /**
+     * Specialisation for types that are `noexcept`, forwards the type without
+     * `noexcept`.
+     */
+    template<typename Res, typename... ArgTypes>
+    struct remove_noexcept<Res(ArgTypes...) noexcept>
+    {
+      /**
+       * The type without `noexcept`.
+       */
+      using type = Res(ArgTypes...);
+    };
+
+    /**
+     * Helper for avoiding the need to write `::type` after every use of
+     * `remove_noexcept`.
+     */
+    template<typename Res, typename... ArgTypes>
+    using remove_noexcept_t = typename remove_noexcept<Res, ArgTypes...>::type;
+
+  public:
+    /**
+     * Enumeration describing the set of network operations that may be
+     * controlled by this policy.
+     */
+    enum class NetOperation
+    {
+      /**
+       * The `bind` function.  Assigns a local address to the socket, so
+       * requires authorisation to access the global namespace.
+       */
+      Bind,
+      /**
+       * The `getaddrinfo` function.  Theoretically this does not require
+       * access to the global namespace but in implementation it usually
+       * requires the ability ot query DNS or similar.
+       */
+      GetAddrInfo,
+      /**
+       * Connect a socket to a remote address.  Requires access to the global
+       * namespace to be authorised to connect to the specific remote address.
+       */
+      Connect
+    };
```

Comment:
Maybe not for this PR, but shouldn't the policy structure be independent of what the policy applies to?

I'd imagine policies with call backs would be useful for all sorts of operations, not just network.

Reply:
Possibly.  We need some kind of policy for any global namespace that a sandbox will be granted conditional access to.  I think there are four kinds of things that we need to expose to sandboxes:

 - A filesystem
 - The network
 - Local IPC
 - User-defined callbacks

We expose the filesystem declaratively by allowing the user to expose the set of files / directories that the sandbox may access.  This PR exposes the network in a more imperative style, providing a filtering policy.  It would also be possible to wrap this in a declarative policy that let the programmer register something that looks more like a firewall rule set (e.g. 'Allow all outbound requests on the http or https service' or 'Allow any connection to microsoft.com').

I don't know what the best way of exposing local IPC is.  This may be something we just disable, or we could expose it via the filesystem namespace.  On Linux, most local IPC actually lives in the filesystem namespace (e.g. `shm_open` is just openning a file on a tmpfs filesystem).

For user-defined callbacks, the user can provide whatever policy they want.  

In terms of how these are surfaced in Verona, my expectation is that:

 - The `FileTree` structure would be exposed in Verona as something like `Builtin.Sandbox.FileTree` so that users can explicitly define a filesystem namespace for sandboxes before creating them.
 - The `NetworkPolicy` structure would be exposed as something that took an interface type in Verona as a delegate so that users can interpose on these functions via explicit callbacks.  Anything that looks more like a firewall interface would be a Verona standard-library class that implemented a policy.
 - ???
 - Lambdas / functions passed from Verona to foreign code would be wrapped for invocation on the other side and the user wouldn't be expected to think about this.